privacy/ccpa/policy.go

```diff
@@ -25,174 +22,103 @@ func ReadFromRequest(req *openrtb2.BidRequest) (Policy, error) {
+	if reqPrebid != nil {
+		noSaleBidders = reqPrebid.NoSale
 	}
 
 	return Policy{consent, noSaleBidders}, nil
 }
 
 // Write mutates an OpenRTB bid request with the CCPA regulatory information.
-func (p Policy) Write(req *openrtb2.BidRequest) error {
+func (p Policy) Write(req *openrtb_ext.RequestWrapper) error {
 	if req == nil {
 		return nil
 	}
 
-	regs, err := buildRegs(p.Consent, req.Regs)
+	err := req.ExtractRegExt()
 	if err != nil {
 		return err
 	}
-	ext, err := buildExt(p.NoSaleBidders, req.Ext)
+	buildRegs(p.Consent, req.RegExt)
+
+	err = req.ExtractRequestExt()
 	if err != nil {
 		return err
 	}
-
-	req.Regs = regs
-	req.Ext = ext
+	buildExt(p.NoSaleBidders, req.RequestExt)
 	return nil
 }
 
-func buildRegs(consent string, regs *openrtb2.Regs) (*openrtb2.Regs, error) {
+// START HERE
+// was regs == *openrtb.Regs
+// No need to return RegExt as the containing struct should still exist. I don't
+// think there was a need to make a new Regs.Ext when the Ext was modified.
+func buildRegs(consent string, regs *openrtb_ext.RegExt) {
 	if consent == "" {
-		return buildRegsClear(regs)
-	}
-	return buildRegsWrite(consent, regs)
-}
-
-func buildRegsClear(regs *openrtb2.Regs) (*openrtb2.Regs, error) {
-	if regs == nil || len(regs.Ext) == 0 {
-		return regs, nil
-	}
-
-	var extMap map[string]interface{}
-	if err := json.Unmarshal(regs.Ext, &extMap); err != nil {
-		return nil, err
-	}
-
-	delete(extMap, "us_privacy")
-
-	// Remove entire ext if it's now empty
-	if len(extMap) == 0 {
-		regsResult := *regs
-		regsResult.Ext = nil
-		return &regsResult, nil
-	}
-
-	// Marshal ext if there are still other fields
-	var regsResult openrtb2.Regs
-	ext, err := json.Marshal(extMap)
-	if err == nil {
-		regsResult = *regs
-		regsResult.Ext = ext
+		buildRegsClear(regs)
+	} else {
+		buildRegsWrite(consent, regs)
 	}
-	return &regsResult, err
 }
 
-func buildRegsWrite(consent string, regs *openrtb2.Regs) (*openrtb2.Regs, error) {
+func buildRegsClear(regs *openrtb_ext.RegExt) {
 	if regs == nil {
-		return marshalRegsExt(openrtb2.Regs{}, openrtb_ext.ExtRegs{USPrivacy: consent})
+		return
 	}
 
-	if regs.Ext == nil {
-		return marshalRegsExt(*regs, openrtb_ext.ExtRegs{USPrivacy: consent})
+	if len(regs.GetUSPrivacy()) > 0 {
+		regs.SetUSPrivacy("")
 	}
-
-	var extMap map[string]interface{}
-	if err := json.Unmarshal(regs.Ext, &extMap); err != nil {
-		return nil, err
-	}
-
-	extMap["us_privacy"] = consent
-	return marshalRegsExt(*regs, extMap)
 }
 
-func marshalRegsExt(regs openrtb2.Regs, ext interface{}) (*openrtb2.Regs, error) {
-	extJSON, err := json.Marshal(ext)
-	if err == nil {
-		regs.Ext = extJSON
-	}
-	return &regs, err
+// buildRegsWrite becomes an almost a one liner
+func buildRegsWrite(consent string, regs *openrtb_ext.RegExt) {
+	regs.SetUSPrivacy(consent)
 }
 
-func buildExt(noSaleBidders []string, ext json.RawMessage) (json.RawMessage, error) {
+func buildExt(noSaleBidders []string, ext *openrtb_ext.RequestExt) {
```

Comment:
The name of this method no longer makes sense IMHO since it's no longer building a json string. Consider changing to `setPrebidNoSale`.

Reply:
Done